// angular2-notifications/esm5/lib/components/notification/notification.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { animate, state, style, transition, trigger } from '@angular/animations';
import { ChangeDetectionStrategy, Component, Input, NgZone, OnDestroy, OnInit, TemplateRef, ViewEncapsulation, ChangeDetectorRef, ViewRef } from '@angular/core';
import { DomSanitizer, SafeHtml } from '@angular/platform-browser';
import { NotificationsService } from '../../services/notifications.service';
var NotificationComponent = /** @class */ (function () {
    function NotificationComponent(notificationService, domSanitizer, cd, zone) {
        var _this = this;
        this.notificationService = notificationService;
        this.domSanitizer = domSanitizer;
        this.cd = cd;
        this.zone = zone;
        this.titleIsTemplate = false;
        this.contentIsTemplate = false;
        this.htmlIsTemplate = false;
        this.progressWidth = 0;
        this.stopTime = false;
        this.framesPerSecond = 40;
        this.instance = function () {
            var now = new Date().getTime();
            if (_this.endTime < now) {
                _this.remove();
                _this.item.timeoutEnd.emit();
            }
            else if (!_this.stopTime) {
                if (_this.showProgressBar) {
                    // We add this.sleepTime just to have 100% before close
                    _this.progressWidth = Math.min((now - _this.startTime + _this.sleepTime) * 100 / _this.timeOut, 100);
                }
                _this.timer = setTimeout(_this.instance, _this.sleepTime);
            }
            _this.zone.run(function () {
                if (!_this.cd.destroyed) {
                    _this.cd.detectChanges();
                }
            });
        };
    }
    NotificationComponent.prototype.ngOnInit = function () {
        if (this.item.override) {
            this.attachOverrides();
        }
        if (this.animate) {
            this.item.state = this.animate;
        }
        if (this.timeOut !== 0) {
            this.startTimeOut();
        }
        this.contentType(this.item.title, 'title');
        this.contentType(this.item.content, 'content');
        this.contentType(this.item.html, 'html');
        this.safeSvg = this.domSanitizer.bypassSecurityTrustHtml(this.icon || this.item.icon);
        this.safeInputHtml = this.domSanitizer.bypassSecurityTrustHtml(this.item.html);
    };
    NotificationComponent.prototype.ngOnDestroy = function () {
        clearTimeout(this.timer);
        this.cd.detach();
    };
    NotificationComponent.prototype.startTimeOut = function () {
        var _this = this;
        this.sleepTime = 1000 / this.framesPerSecond /* ms */;
        this.startTime = new Date().getTime();
        this.endTime = this.startTime + this.timeOut;
        this.zone.runOutsideAngular(function () { return _this.timer = setTimeout(_this.instance, _this.sleepTime); });
    };
    NotificationComponent.prototype.onEnter = function () {
        if (this.pauseOnHover) {
            this.stopTime = true;
            this.pauseStart = new Date().getTime();
        }
    };
    NotificationComponent.prototype.onLeave = function () {
        var _this = this;
        if (this.pauseOnHover) {
            this.stopTime = false;
            this.startTime += (new Date().getTime() - this.pauseStart);
            this.endTime += (new Date().getTime() - this.pauseStart);
            this.zone.runOutsideAngular(function () { return setTimeout(_this.instance, _this.sleepTime); });
        }
    };
    NotificationComponent.prototype.onClick = function (event) {
        this.item.click.emit(event);
        if (this.clickToClose) {
            this.remove();
        }
    };
    NotificationComponent.prototype.onClickIcon = function (event) {
        this.item.clickIcon.emit(event);
        if (this.clickIconToClose) {
            this.remove();
        }
    };
    // Attach all the overrides
    NotificationComponent.prototype.attachOverrides = function () {
        var _this = this;
        Object.keys(this.item.override).forEach(function (a) {
            if (_this.hasOwnProperty(a)) {
                _this[a] = _this.item.override[a];
            }
        });
    };
    NotificationComponent.prototype.remove = function () {
        var _this = this;
        if (this.animate) {
            this.item.state = this.animate + 'Out';
            setTimeout(function () {
                _this.notificationService.set(_this.item, false);
            }, 310);
        }
        else {
            this.notificationService.set(this.item, false);
        }
    };
    NotificationComponent.prototype.contentType = function (item, key) {
        if (item instanceof TemplateRef) {
            this[key] = item;
        }
        else {
            this[key] = this.domSanitizer.bypassSecurityTrustHtml(item);
        }
        this[key + 'IsTemplate'] = item instanceof TemplateRef;
    };
    NotificationComponent.ctorParameters = function () { return [
        { type: NotificationsService },
        { type: DomSanitizer },
        { type: ChangeDetectorRef },
        { type: NgZone }
    ]; };
    __decorate([
        Input()
    ], NotificationComponent.prototype, "timeOut", void 0);
    __decorate([
        Input()
    ], NotificationComponent.prototype, "showProgressBar", void 0);
    __decorate([
        Input()
    ], NotificationComponent.prototype, "pauseOnHover", void 0);
    __decorate([
        Input()
    ], NotificationComponent.prototype, "clickToClose", void 0);
    __decorate([
        Input()
    ], NotificationComponent.prototype, "clickIconToClose", void 0);
    __decorate([
        Input()
    ], NotificationComponent.prototype, "maxLength", void 0);
    __decorate([
        Input()
    ], NotificationComponent.prototype, "theClass", void 0);
    __decorate([
        Input()
    ], NotificationComponent.prototype, "rtl", void 0);
    __decorate([
        Input()
    ], NotificationComponent.prototype, "animate", void 0);
    __decorate([
        Input()
    ], NotificationComponent.prototype, "position", void 0);
    __decorate([
        Input()
    ], NotificationComponent.prototype, "item", void 0);
    NotificationComponent = __decorate([
        Component({
            selector: 'simple-notification',
            encapsulation: ViewEncapsulation.None,
            animations: [
                trigger('enterLeave', [
                    // Fade
                    state('fade', style({ opacity: 1 })),
                    transition('* => fade', [
                        style({ opacity: 0 }),
                        animate('400ms ease-in-out')
                    ]),
                    state('fadeOut', style({ opacity: 0 })),
                    transition('fade => fadeOut', [
                        style({ opacity: 1 }),
                        animate('300ms ease-in-out')
                    ]),
                    // Enter from top
                    state('fromTop', style({ opacity: 1, transform: 'translateY(0)' })),
                    transition('* => fromTop', [
                        style({ opacity: 0, transform: 'translateY(-5%)' }),
                        animate('400ms ease-in-out')
                    ]),
                    state('fromTopOut', style({ opacity: 0, transform: 'translateY(5%)' })),
                    transition('fromTop => fromTopOut', [
                        style({ opacity: 1, transform: 'translateY(0)' }),
                        animate('300ms ease-in-out')
                    ]),
                    // Enter from right
                    state('fromRight', style({ opacity: 1, transform: 'translateX(0)' })),
                    transition('* => fromRight', [
                        style({ opacity: 0, transform: 'translateX(5%)' }),
                        animate('400ms ease-in-out')
                    ]),
                    state('fromRightOut', style({ opacity: 0, transform: 'translateX(-5%)' })),
                    transition('fromRight => fromRightOut', [
                        style({ opacity: 1, transform: 'translateX(0)' }),
                        animate('300ms ease-in-out')
                    ]),
                    // Enter from bottom
                    state('fromBottom', style({ opacity: 1, transform: 'translateY(0)' })),
                    transition('* => fromBottom', [
                        style({ opacity: 0, transform: 'translateY(5%)' }),
                        animate('400ms ease-in-out')
                    ]),
                    state('fromBottomOut', style({ opacity: 0, transform: 'translateY(-5%)' })),
                    transition('fromBottom => fromBottomOut', [
                        style({ opacity: 1, transform: 'translateY(0)' }),
                        animate('300ms ease-in-out')
                    ]),
                    // Enter from left
                    state('fromLeft', style({ opacity: 1, transform: 'translateX(0)' })),
                    transition('* => fromLeft', [
                        style({ opacity: 0, transform: 'translateX(-5%)' }),
                        animate('400ms ease-in-out')
                    ]),
                    state('fromLeftOut', style({ opacity: 0, transform: 'translateX(5%)' })),
                    transition('fromLeft => fromLeftOut', [
                        style({ opacity: 1, transform: 'translateX(0)' }),
                        animate('300ms ease-in-out')
                    ]),
                    // Rotate
                    state('scale', style({ opacity: 1, transform: 'scale(1)' })),
                    transition('* => scale', [
                        style({ opacity: 0, transform: 'scale(0)' }),
                        animate('400ms ease-in-out')
                    ]),
                    state('scaleOut', style({ opacity: 0, transform: 'scale(0)' })),
                    transition('scale => scaleOut', [
                        style({ opacity: 1, transform: 'scale(1)' }),
                        animate('400ms ease-in-out')
                    ]),
                    // Scale
                    state('rotate', style({ opacity: 1, transform: 'rotate(0deg)' })),
                    transition('* => rotate', [
                        style({ opacity: 0, transform: 'rotate(5deg)' }),
                        animate('400ms ease-in-out')
                    ]),
                    state('rotateOut', style({ opacity: 0, transform: 'rotate(-5deg)' })),
                    transition('rotate => rotateOut', [
                        style({ opacity: 1, transform: 'rotate(0deg)' }),
                        animate('400ms ease-in-out')
                    ])
                ])
            ],
            template: "<div class=\"simple-notification\"\r\n     [@enterLeave]=\"item.state\"\r\n     (click)=\"onClick($event)\"\r\n     [class]=\"theClass\"\r\n     [ngClass]=\"{\r\n            'alert': item.type === 'alert',\r\n            'error': item.type === 'error',\r\n            'warn': item.type === 'warn',\r\n            'success': item.type === 'success',\r\n            'info': item.type === 'info',\r\n            'bare': item.type === 'bare',\r\n            'rtl-mode': rtl,\r\n            'has-icon': item.icon !== 'bare'\r\n        }\"\r\n     (mouseenter)=\"onEnter()\"\r\n     (mouseleave)=\"onLeave()\">\r\n\r\n    <div *ngIf=\"!item.html\">\r\n\r\n        <div class=\"sn-title\" *ngIf=\"titleIsTemplate; else regularTitle\">\r\n            <ng-container *ngTemplateOutlet=\"title; context: item.context\"></ng-container>\r\n        </div>\r\n\r\n        <ng-template #regularTitle>\r\n            <div class=\"sn-title\" [innerHTML]=\"title\"></div>\r\n        </ng-template>\r\n\r\n        <div class=\"sn-content\" *ngIf=\"contentIsTemplate else regularContent\">\r\n            <ng-container *ngTemplateOutlet=\"content; context: item.context\"></ng-container>\r\n        </div>\r\n\r\n        <ng-template #regularContent>\r\n            <div class=\"sn-content\" [innerHTML]=\"content\"></div>\r\n        </ng-template>\r\n\r\n        <div class=\"icon\" *ngIf=\"item.icon !== 'bare'\" [innerHTML]=\"safeSvg\"></div>\r\n    </div>\r\n    <div *ngIf=\"item.html\">\r\n        <div class=\"sn-html\" *ngIf=\"htmlIsTemplate; else regularHtml\">\r\n            <ng-container *ngTemplateOutlet=\"item.html; context: item.context\"></ng-container>\r\n        </div>\r\n\r\n        <ng-template #regularHtml>\r\n            <div class=\"sn-content\" [innerHTML]=\"safeInputHtml\"></div>\r\n        </ng-template>\r\n\r\n        <div class=\"icon\" [class.icon-hover]=\"clickIconToClose\" *ngIf=\"item.icon\" [innerHTML]=\"safeSvg\" (click)=\"onClickIcon($event)\"></div>\r\n    </div>\r\n\r\n    <div class=\"sn-progress-loader\" *ngIf=\"showProgressBar\">\r\n        <span [ngStyle]=\"{'width': progressWidth + '%'}\"></span>\r\n    </div>\r\n\r\n</div>\r\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [".simple-notification{width:100%;padding:10px 20px;box-sizing:border-box;position:relative;float:left;margin-bottom:10px;color:#fff;cursor:pointer;transition:.5s;min-height:70px}.simple-notification .sn-content,.simple-notification .sn-html,.simple-notification .sn-title{margin:0}.simple-notification .sn-title{line-height:30px;font-size:20px}.simple-notification .sn-content{font-size:16px;line-height:20px}.simple-notification.has-icon .sn-content,.simple-notification.has-icon .sn-html,.simple-notification.has-icon .sn-title{padding:0 50px 0 0}.simple-notification .icon{position:absolute;box-sizing:border-box;top:0;right:0;width:70px;height:70px;padding:10px}.simple-notification .icon.icon-hover:hover{opacity:.5}.simple-notification .icon svg{fill:#fff;width:100%;height:100%}.simple-notification .icon svg g{fill:#fff}.simple-notification.rtl-mode.has-icon .sn-content,.simple-notification.rtl-mode.has-icon .sn-html,.simple-notification.rtl-mode.has-icon .sn-title{padding:0 0 0 50px}.simple-notification.rtl-mode{direction:rtl}.simple-notification.rtl-mode .sn-content{padding:0 0 0 50px}.simple-notification.rtl-mode svg{left:0;right:auto}.simple-notification.error{background:#f44336}.simple-notification.success{background:#8bc34a}.simple-notification.alert{background:#ffdb5b}.simple-notification.info{background:#03a9f4}.simple-notification.warn{background:#ffdb5b}.simple-notification .sn-progress-loader{position:absolute;top:0;left:0;width:100%;height:5px}.simple-notification .sn-progress-loader span{float:left;height:100%}.simple-notification.success .sn-progress-loader span{background:#689f38}.simple-notification.error .sn-progress-loader span{background:#d32f2f}.simple-notification.alert .sn-progress-loader span{background:#edc242}.simple-notification.info .sn-progress-loader span{background:#0288d1}.simple-notification.warn .sn-progress-loader span{background:#edc242}.simple-notification.bare .sn-progress-loader span{background:#ccc}.simple-notification.warn div .sn-content,.simple-notification.warn div .sn-html,.simple-notification.warn div .sn-title{color:#444}"]
        })
    ], NotificationComponent);
    return NotificationComponent;
}());
export { NotificationComponent };
